File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/domainservice/LegoStockShipTest.java
package eu.luminis.workshop.smallsteps.logic.domainservice;

import eu.luminis.workshop.smallsteps.logic.domainservice.auth.AccessDeniedException;
import eu.luminis.workshop.smallsteps.logic.domainservice.helper.SetupLegoTestApp;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoSetForBuilder;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;
import eu.luminis.workshop.smallsteps.persistence.InMemoryLegoPartCatalog;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class LegoStockShipTest {
    SetupLegoTestApp app;

    @BeforeEach
    void setUp() {
        app = new SetupLegoTestApp();
    }

    @Test
    public void should_register_shipping_lego_set_when_reserved() {
        StockState initialState = new StockState(
                null,
                Set.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null,
                null
        );
        LegoStock handler = new LegoStock(app.bussumAuth, new InMemoryLegoPartCatalog(), initialState);

        StockState newState = handler.ship(app.millenniumFalcon, app.harry);

        assertThat(newState.getReserved()).isEmpty();
        assertThat(newState.getAtBuilder()).isEqualTo(List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)));
        assertThat(newState.getAvailableStock()).isEqualTo(initialState.getAvailableStock());
        assertThat(newState.getIncompleteStock()).isEqualTo(initialState.getIncompleteStock());
        assertThat(newState.getIncompleteReturnHistory()).isEqualTo(initialState.getIncompleteReturnHistory());
    }

    @Test
    public void should_fail_to_ship_when_not_reserved_by_builder() {
        StockState initialState = new StockState(
                null,
                Collections.emptySet(),
                null,
                null,
                null
        );
        LegoStock handler = new LegoStock(app.bussumAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            handler.ship(app.millenniumFalcon, app.harry);
        }).isInstanceOf(IllegalStateException.class)
                .hasMessage("Legoset " + app.millenniumFalcon.getNumber() + " has not been reserved by builder " + app.harry.toString());

    }

    @Test
    public void should_fail_to_ship_when_not_authenticated_as_shop() {
        StockState initialState = new StockState(
                null,
                Set.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            handler.ship(app.millenniumFalcon, app.harry);
        }).isInstanceOf(AccessDeniedException.class);
    }
}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/domainService/LegoStock - ShipTest.kt
package eu.luminis.workshop.smallsteps.logic.domainService

import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper.millenniumFalcon
import eu.luminis.workshop.smallsteps.logic.domainService.auth.AccessDeniedException
import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper
import eu.luminis.workshop.smallsteps.logic.domainService.helper.TestUsers
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

@Suppress("ClassName")
internal class `LegoStock - ShipTest` {
    @Test
    fun `should register shipping lego set when reserved`() {
        // given
        val initialState = StockState(
            reserved = setOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(initialState, auth = TestUsers.bussumAuth)

        // when
        val newState = handler.ship(millenniumFalcon, TestUsers.harry)

        // then
        assertEquals(emptySet(), newState.reserved)
        assertEquals(listOf(millenniumFalcon to TestUsers.harry), newState.atBuilder)
        assertEquals(initialState.availableStock, newState.availableStock)
        assertEquals(initialState.incompleteStock, newState.incompleteStock)
        assertEquals(initialState.incompleteReturnHistory, newState.incompleteReturnHistory)
    }

    @Test
    fun `should fail to ship when not reserved by builder`() {
        // given
        val initialState = StockState(
            reserved = emptySet(),
        )
        val handler = LegoStockHelper.build(initialState, auth = TestUsers.bussumAuth)

        // then
        assertFailsWith<IllegalStateException> {
            // when
            handler.ship(millenniumFalcon, TestUsers.harry)
        }
    }

    @Test
    fun `should fail to ship when not authenticated as shop`() {
        // given
        val initialState = StockState(
            reserved = setOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(initialState, auth = TestUsers.harryAuth)

        // then
        assertFailsWith<AccessDeniedException> {
            // when
            handler.ship(millenniumFalcon, TestUsers.harry)
        }
    }
}